Sources/MlemMiddleware/FeedLoaders/User Content Feed Loader/UserContentFeedLoader.swift

```diff
@@ -0,0 +1,240 @@
+    
+    private func loadContentPage(_ pageToLoad: Int) async throws {
+        await contentLoadingSemaphore.wait()
+        defer { contentLoadingSemaphore.signal() }
+        
+        loadingState = .loading
+        
+        guard pageToLoad == contentPage + 1 else {
+            print("Unexpected content page \(pageToLoad) encountered (expected \(contentPage + 1), skipping load")
+            return
+        }
+        
+        contentPage += 1
+        var newItems: [UserContent] = .init()
+        while newItems.count < 50, loadingState != .done {
+            if let nextItem = try await computeNextItem() {
+                newItems.append(nextItem)
+            } else {
+                loadingState = .done
+            }
+        }
+        
+        if pageToLoad == 1 {
+            await setItems(newItems)
+        } else {
+            await addItems(newItems)
+        }
+        
+        if loadingState != .done {
+            loadingState = .idle
+            thresholds.update(with: newItems)
+        }
+    }
+    
+    /// Returns the next post or comment, depending on which is sorted first
+    private func computeNextItem() async throws -> UserContent? {
+        // if either postStream or commentStream needs items, load
+        if postStream.needsMoreItems || commentStream.needsMoreItems {
+            try await loadNextApiPage()
+        }
+        
+        let nextPost = try await postStream.nextItemSortVal(sortType: sortType)
+        let nextComment = try await commentStream.nextItemSortVal(sortType: sortType)
+        
+        if let nextPost {
+            if let nextComment {
+                // if both next post and next comment, return higher sort
+                return nextPost > nextComment ? postStream.consumeNextItem() : commentStream.consumeNextItem()
+            } else {
+                // if next post but no next comment, return next post
+                return postStream.consumeNextItem()
+            }
+        }
+        
+        // if no next post, always return next comment (this returns nil if no next comment)
+        return commentStream.consumeNextItem()
+    }
+    
+    // MARK: Helpers
+    
+    @MainActor
+    private func addItems(_ newItems: [UserContent]) {
+        items.append(contentsOf: newItems)
+    }
+    
+    @MainActor
+    private func setItems(_ newItems: [UserContent]) {
+        items = newItems
+    }
+    
+    private func fetchItems() async throws -> (posts: [Post2], comments: [Comment2]) {
+        return try await api.getContent(authorId: userId, sort: .new, page: apiPage, limit: 50, savedOnly: savedOnly)
+    }
+    
+    /// Loads the next page from the API
+    /// - Warning: This is NOT a thread-safe function! It should only be called from a concurrency-controlled environment
+    private func loadNextApiPage() async throws {
+        apiPage += 1
+        let response = try await fetchItems()
+        preloadImages(response.posts) // TODO: comment images?
```

Comment:
> `// TODO: comment images?`

I'll be adding a way to extract images from a markdown tree at some point (issue [here](https://github.com/mlemgroup/LemmyMarkdownUI/issues/17)).

Reply:
Nice, I've opened #36 to circle back to this when that's done